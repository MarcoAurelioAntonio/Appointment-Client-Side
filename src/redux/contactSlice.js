import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

export const postContactForm = createAsyncThunk(
  'contact/postContactForm',
  async (requestForm) => {
    const formData = new FormData();
    formData.append('contact[name]', requestForm.name);
    formData.append('contact[email]', requestForm.email);
    formData.append('contact[message]', requestForm.message);

    const postFormToData = await axios.post('http://localhost:3000/api/v1/contacts', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return postFormToData.data;
  },
);

const contactSlice = createSlice({
  name: 'contact',
  initialState: {
    contact: [],
    status: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(postContactForm.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(postContactForm.fulfilled, (state, { payload }) => ({
        ...state,
        contact: payload.data,
        status: 'succeeded',
      }))
      .addCase(postContactForm.rejected, (state, { error }) => ({
        ...state,
        status: 'failed',
        error: error.message,
      }));
  },
});

export default contactSlice.reducer;
