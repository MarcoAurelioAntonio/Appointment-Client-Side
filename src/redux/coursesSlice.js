import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

export const getAllCoursesApi = createAsyncThunk('games/fetch', () => (
  new Promise((resolve, reject) => {
    axios.get('http://localhost:3000/api/v1/courses')
      .then(({ data }) => {
        // console.log(data);
        resolve({ data });
      })
      .catch((error) => {
        reject(error);
      });
  })
));

export const postApiCourseForm = createAsyncThunk(
  'courses/postApiCourseForm',
  async (requestForm) => {
    const postFormToData = await axios.post('http://localhost:3000/api/v1/courses', requestForm);
    // console.log(postFormToData.data);
    return postFormToData.data;
  },
);

export const getCourseById = createAsyncThunk(
  'courses/getCourseById',
  async (id) => {
    const response = await axios.get(`http://localhost:3000/api/v1/courses/${id}`);
    return response.data;
  },
);

export const deleteCourseById = createAsyncThunk(
  'courses/deleteCourseById',
  async (id) => {
    const response = await axios.delete(`http://localhost:3000/api/v1/courses/${id}`);
    const payload = { id, response };
    return payload;
  },
);

const coursesSlice = createSlice({
  name: 'courses',
  initialState: {
    courses: [],
    course: {},
    status: 'idle',
    loading: false,
    error: null,
    deleted: false,
  },
  reducers: {
  },
  extraReducers: (builder) => {
    builder
      // getAllCoursesApi
      .addCase(getAllCoursesApi.pending, (state) => ({
        ...state, status: 'loading', loading: true, courses: [],
      }))
      .addCase(getAllCoursesApi.fulfilled, (state, { payload }) => ({
        ...state,
        courses: payload.data,
        status: 'succeeded',
        loading: false,
      }))
      .addCase(getAllCoursesApi.rejected, (state, { error }) => ({
        ...state,
        error: error.message,
        status: 'failed',
      }))

      // postApiCourseForm
      .addCase(postApiCourseForm.pending, (state) => ({
        ...state, status: 'loading', courses: [],
      }))
      .addCase(postApiCourseForm.fulfilled, (state, { payload }) => ({
        ...state,
        courses: payload.array,
        status: 'succeed',
      }))
      .addCase(postApiCourseForm.rejected, (state, { error }) => ({
        ...state,
        error: error.message,
        status: 'failed',
        loading: false,
      }))
      .addCase(getCourseById.pending, (state) => ({
        ...state, status: 'loading', course: {},
      }))
      .addCase(getCourseById.fulfilled, (state, { payload }) => ({
        ...state, loading: false, course: payload, status: 'succeed',
      }))
      .addCase(getCourseById.rejected, (state, { error }) => ({
        ...state,
        error: error.message,
        status: 'failed',
      }))
      .addCase(deleteCourseById.pending, (state) => ({
        ...state,
      }))
      .addCase(deleteCourseById.fulfilled, (state, { payload }) => {
        const { id } = payload;
        const newCourses = state.courses.filter((course) => course.id !== id);
        return {
          ...state, loading: false, courses: newCourses, deleted: true,
        };
      })
      .addCase(deleteCourseById.rejected, (state, { error }) => ({
        ...state, status: 'failed', error: error.message,
      }));
  },
});

export default coursesSlice.reducer;
