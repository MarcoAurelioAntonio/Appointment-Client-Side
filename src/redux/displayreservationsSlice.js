import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

export const getAllReservationsApi = createAsyncThunk(
  'api/getAllReservationsApi',
  async (userId) => {
    const response = await axios.get(`http://localhost:3000/api/v1/users/${userId}/reservations`);
    // console.log(response.data);
    return response.data;
  },
);

const displayreservationsSlice = createSlice({
  name: 'api',
  initialState: {
    data: [],
    status: null,
    error: null,
    userId: null,
  },

  reducers: {
    setUserId: (state, action) => {
      state.userId = action.payload;
    },
  },

  extraReducers: (builder) => {
    builder
      .addCase(getAllReservationsApi.pending, (state) => ({ ...state, status: 'loading' }))
      .addCase(getAllReservationsApi.fulfilled, (state, action) => ({ ...state, status: 'succeeded', data: action.payload }))
      .addCase(getAllReservationsApi.rejected, (state, action) => ({ ...state, status: 'failed', error: action.error.message }));
  },
});

export default displayreservationsSlice.reducer;
