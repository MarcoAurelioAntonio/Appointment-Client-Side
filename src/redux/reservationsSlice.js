import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

export const getAllReservationsApi = createAsyncThunk(
  'api/getAllReservationsApi',
  async (userId) => {
    const response = await axios.get(`http://localhost:3000/api/v1/users/${userId}/reservations`);
    return response.data;
  },
);

export const postReservationToAPI = createAsyncThunk(
  'reservations/postReservationToAPI',
  async (data, thunkAPI) => {
    try {
      const response = await axios.post(
        `http://localhost:3000/api/v1/users/${data.user_id}/reservations`,
        data,
      );
      return response.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.response.data.error);
    }
  },
);
const reservationsSlice = createSlice({
  name: 'api',
  initialState: {
    reservations: [],
    status: 'idle',
    loading: false,
    message: '',
    error: null,
    userId: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(postReservationToAPI.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(postReservationToAPI.fulfilled, (state, { payload }) => ({
        ...state,
        status: 'succeed',
        message: payload.data,
      }))
      .addCase(postReservationToAPI.rejected, (state, action) => ({
        ...state,
        error: action.payload,
        status: 'failed',
      }))
      .addCase(getAllReservationsApi.pending, (state) => ({ ...state, status: 'loading' }))
      .addCase(getAllReservationsApi.fulfilled, (state, action) => ({ ...state, status: 'succeeded', reservations: action.payload }))
      .addCase(getAllReservationsApi.rejected, (state, action) => ({ ...state, status: 'failed', error: action.error.message }));
  },
});

export default reservationsSlice.reducer;
