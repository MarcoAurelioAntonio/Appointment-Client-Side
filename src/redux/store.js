import { configureStore } from '@reduxjs/toolkit';
import coursesReducer from './coursesSlice';
import reservationsReducer from './reservationsSlice';
import displayreservationsReducer from './displayreservationsSlice';

const store = configureStore({
  reducer: {
    courses: coursesReducer,
    reservations: reservationsReducer,
    displayreservations: displayreservationsReducer,
  },
});

export default store;
