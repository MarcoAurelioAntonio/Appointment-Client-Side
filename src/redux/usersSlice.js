import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

export const postUserToAPI = createAsyncThunk(
  'users/postUserToAPI',
  async (data, thunkAPI) => {
    try {
      const response = await axios.post('http://localhost:3000/api/v1/users', data);
      console.log(data);
      return response.data;
    } catch (error) {
      console.log(error);
      return thunkAPI.rejectWithValue(error);
    }
  },
);

const usersSlice = createSlice({
  name: 'users',
  initialState: {
    current_user_id: null,
    status: 'idle',
    loading: false,
    error: null,
  },
  reducers: {
  },
  extraReducers: (builder) => {
    builder
      .addCase(postUserToAPI.pending, (state) => ({
        ...state, status: 'loading',
      }))
      .addCase(postUserToAPI.fulfilled, (state, { payload }) => ({
        ...state,
        current_user_id: payload.id,
        status: 'succeed',
      }))
      .addCase(postUserToAPI.rejected, (state, { error }) => ({
        ...state,
        error: error.message,
        status: 'failed',
      }));
  },
});

export default usersSlice.reducer;
