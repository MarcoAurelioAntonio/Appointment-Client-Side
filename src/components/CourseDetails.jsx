import React, { useEffect } from 'react';
import { useParams, Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getCourseById } from '../redux/coursesSlice';
import NavMenu from './NavMenu';

const CourseDetails = () => {
  const { id } = useParams();
  const { course } = useSelector((state) => state.courses);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getCourseById(id));
  }, [dispatch, id]);

  return (
    <div className="flex flex-col md:flex-row justify-between my-16 h-screen gap-12 px-10">

      <NavMenu bgColor="" isHide isBacking />
      <div className="flex flex-col justify-start items-start w-full">
        <div className="flex flex-col items-center justify-center md:flex-row flex-shrink gap-12 w-full md:items-start">
          <div className="flex flex-col w-1/2 ml-8">
            <img
              className="course-image"
              width="100%"
              height="50%"
              src={course.image_url}
              alt={course.title}
            />
          </div>

          <div className="w-full px-5 flex flex-col md:w-1/2">
            <h1 className="text-4xl text-end">{course.name}</h1>
            <p className="text-end">{course.description}</p>
            <ul className="mt-4">
              <li className="flex justify-between border p-3  bg-slate-300 ">
                <p>Course type</p>
                <p className="font-bold">
                  {' '}
                  {course.course_type}
                </p>
              </li>
              <li className="flex justify-between  p-3 ">
                <p>Course price</p>
                <p className="font-bold">
                  {' '}
                  {course.price}
                </p>
              </li>
              <li className="flex justify-between border p-3  bg-slate-300">
                <p>Start date</p>
                <p className="font-bold">{course.start_date}</p>
              </li>
              <li className="flex justify-between p-3 ">
                <p>End date</p>
                <p className="font-bold">
                  {' '}
                  {course.end_date}
                </p>
              </li>
            </ul>

            <Link
              className="enroll-btn"
              to="/add-reservation"
              state={course}
            >
              Enroll Course
              <span className="material-symbols-outlined">
                arrow_circle_right
              </span>
            </Link>
          </div>
        </div>
      </div>

    </div>
  );
};

export default CourseDetails;
