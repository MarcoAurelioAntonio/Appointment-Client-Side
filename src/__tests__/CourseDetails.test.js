import { render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import CourseDetails from '../components/CourseDetails';
import { getCourseById } from '../redux/coursesSlice';

describe('Tests for <CourseDetails /> component', () => {
  const mockStore = configureStore([thunk]);
  const user = {
    id: 1,
    name: 'Teddy',
  };
  const course = {
    id: 1,
    name: 'Master Data Structures',
    description: 'Description related to the course',
    course_type: 'Programming',
    price: '200.0',
    start_date: '2023-06-21',
    end_date: '2023-07-21',
    created_at: '2023-06-20T22:03:45.571Z',
    updated_at: '2023-06-20T22:03:45.861Z',
    image_url: 'http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCZz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2e2c6f24621c425f569dcfdc5f64d41d7aa1f592/pexels-karolina-grabowska-4497759.jpg',
  };
  const mockResponse = {
    users: {
      current_user: user,
      status: 'idle',
      loading: false,
      error: null,
    },
    courses: {
      courses: [],
      course,
      status: 'idle',
      loading: false,
      error: null,
      deleted: false,
    },
  };
  jest.spyOn(global, 'fetch').mockResolvedValue({
    json: jest.fn().mockResolvedValue(mockResponse),
  });
  const store = mockStore(mockResponse); // Initialize the mock store
  store.dispatch(getCourseById(1)); // Dispatch the action
  global.fetch.mockRestore(); // Restore the original `fetch` function
  it('Should display 2 cards on the screen', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <CourseDetails />
          ,
        </BrowserRouter>
      </Provider>
      ,
    );
    const courseName = screen.getByText('Master Data Structures');
    expect(courseName).toBeInTheDocument();
  });
  it('Should match the snapshot', () => {
    expect(renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <CourseDetails />
        </BrowserRouter>
      </Provider>,
    ).toJSON()).toMatchSnapshot();
  });
});
