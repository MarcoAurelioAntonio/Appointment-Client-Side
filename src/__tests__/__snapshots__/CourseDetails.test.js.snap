// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <CourseDetails /> component Should match the snapshot 1`] = `
<div
  className="flex flex-col md:flex-row justify-between my-16 h-screen gap-12 px-10"
>
  <button
    className="burger-button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="material-symbols-outlined hamburger green"
    >
      menu
    </span>
  </button>
  <div
    className="flex flex-col justify-start items-start w-full"
  >
    <div
      className="flex flex-col items-center justify-center md:flex-row flex-shrink gap-12 w-full md:items-start"
    >
      <div
        className="flex flex-col w-1/2 ml-8"
      >
        <img
          className="course-image"
          height="50%"
          src="http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBCZz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2e2c6f24621c425f569dcfdc5f64d41d7aa1f592/pexels-karolina-grabowska-4497759.jpg"
          width="100%"
        />
      </div>
      <div
        className="w-full px-5 flex flex-col md:w-1/2"
      >
        <h1
          className="text-4xl text-end"
        >
          Master Data Structures
        </h1>
        <p
          className="text-end"
        >
          Description related to the course
        </p>
        <ul
          className="mt-4"
        >
          <li
            className="flex justify-between border p-3  bg-slate-300 "
          >
            <p>
              Course type
            </p>
            <p
              className="font-bold"
            >
               
              Programming
            </p>
          </li>
          <li
            className="flex justify-between  p-3 "
          >
            <p>
              Course price
            </p>
            <p
              className="font-bold"
            >
               
              200.0
            </p>
          </li>
          <li
            className="flex justify-between border p-3  bg-slate-300"
          >
            <p>
              Start date
            </p>
            <p
              className="font-bold"
            >
              2023-06-21
            </p>
          </li>
          <li
            className="flex justify-between p-3 "
          >
            <p>
              End date
            </p>
            <p
              className="font-bold"
            >
               
              2023-07-21
            </p>
          </li>
        </ul>
        <a
          className="enroll-btn"
          href="/add-reservation"
          onClick={[Function]}
        >
          Enroll Course
          <span
            className="material-symbols-outlined"
          >
            arrow_circle_right
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;
